lib/auth/recordingmetadata/recordingmetadatav1/service.go

```diff
@@ -0,0 +1,144 @@
+)
+
+// Service implements the RecordingMetadataServiceServer interface, providing methods to retrieve session recording
+// metadata and thumbnails.
+type Service struct {
+	pb.UnimplementedRecordingMetadataServiceServer
+
+	authorizer    Authorizer
+	streamer      player.Streamer
+	uploadHandler events.UploadHandler
+	logger        *slog.Logger
+}
+
+// Authorizer is an interface that defines the method for authorizing access to session recordings.
+type Authorizer interface {
+	// Authorize checks if the user has permission to access the session recording, called with the session ID.
+	Authorize(context.Context, string) error
+}
+
+// ServiceConfig holds the configuration for the recording metadata service.
+type ServiceConfig struct {
+	// Authorizer is used to check if the user has permission to access the session recording.
+	Authorizer Authorizer
+	// Streamer is used to stream session recordings.
+	Streamer player.Streamer
+	// UploadHandler is used to handle uploads and downloads of session recording metadata and thumbnails.
+	UploadHandler events.UploadHandler
+}
+
+var _ pb.RecordingMetadataServiceServer = (*Service)(nil)
+
+// NewService creates a new instance of the recording metadata service.
+func NewService(cfg ServiceConfig) (*Service, error) {
+	if cfg.Authorizer == nil {
+		return nil, trace.BadParameter("authorizer is required")
+	}
+	if cfg.UploadHandler == nil {
+		return nil, trace.BadParameter("upload handler is required")
+	}
+
+	return &Service{
+		authorizer:    cfg.Authorizer,
+		streamer:      cfg.Streamer,
+		uploadHandler: cfg.UploadHandler,
+		logger:        slog.With(teleport.ComponentKey, "recording_metadata"),
+	}, nil
+}
+
+// GetThumbnail retrieves the thumbnail for a session recording.
+func (r *Service) GetThumbnail(ctx context.Context, req *pb.GetThumbnailRequest) (*pb.GetThumbnailResponse, error) {
+	if err := r.authorizer.Authorize(ctx, req.SessionId); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	buf := &memBuffer{}
+	err := r.uploadHandler.DownloadThumbnail(ctx, session.ID(req.SessionId), buf)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	thumbnail := &pb.SessionRecordingThumbnail{}
+	err = proto.Unmarshal(buf.Bytes(), thumbnail)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &pb.GetThumbnailResponse{Thumbnail: thumbnail}, nil
+}
+
+// GetMetadata retrieves the metadata for a session recording, streaming it back in chunks (one for metadata and one
+// for each frame).
+func (r *Service) GetMetadata(req *pb.GetMetadataRequest, stream grpc.ServerStreamingServer[pb.GetMetadataResponseChunk]) error {
+	if err := r.authorizer.Authorize(stream.Context(), req.SessionId); err != nil {
+		return trace.Wrap(err)
+	}
+
+	buf := &memBuffer{}
+	err := r.uploadHandler.DownloadMetadata(stream.Context(), session.ID(req.SessionId), buf)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	metadata := &pb.SessionRecordingMetadataWithFrames{}
```

Comment:
question: should we write the message as a single protobuf message or should we write multiple protobuf messages into the file? That would simplify the download and read cycles for larger sessions?

Reply:
Updated it to use protodelim